### ISSUE TITLE
Handle BatchNorm and base_module call

### ISSUE THREAD

**Issue by antoinedemathelin** [scikit-adaptation/skada#252]
Hi everyone,
I open this issue because I noticed that the base_module is called separately on source and target. This can cause undesired behavior when using BatchNorm with the estimation of the batch variance and mean. I propose to concatenate source and target and separate the two after forward pass in base_module

https://github.com/scikit-adaptation/skada/blob/e80e2057933defb24f5d93e198e61f0e881d908a/skada/deep/base.py#L421-L442

**Comment by YanisLalou** [comment]
You're right about the potential undesired behaviour when using `BatchNorm`.
However we can't concatenate `source` and `target` since afterward we need to fetch the `source` and `target` features from the `intermediate_layers` (I agree we should add comments in the code to explain this).
https://github.com/scikit-adaptation/skada/blob/e80e2057933defb24f5d93e198e61f0e881d908a/skada/deep/base.py#L427-L436
https://github.com/scikit-adaptation/skada/blob/e80e2057933defb24f5d93e198e61f0e881d908a/skada/deep/base.py#L438-L447
So if you have other ideas on how to fix this issue, don't hesitate to write them down here.
